toolchain/diagnostics/format_providers.h

```diff
@@ -0,0 +1,135 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_DIAGNOSTICS_FORMAT_PROVIDERS_H_
+#define CARBON_TOOLCHAIN_DIAGNOSTICS_FORMAT_PROVIDERS_H_
+
+#include "common/check.h"
+#include "common/ostream.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+
+namespace Carbon {
+
+// Selects a formatv string based on the value.
+//
+// When used, the true and false outputs are separated by a `|`.
+//
+// For example, `{0:true|false}` would yield standard bool formatting.
+//
+// If needed, the _full_ style string can be wrapped with `'` in order to
+// preserve prefix or suffix whitespace (which is stripped by formatv). For
+// example, `{0:' true | false '}` retains whitespace which would be dropped
+// before `true` and after `false`.
+struct FormatBool {
```

Comment:
For the existing diagnostic helpers, the naming convention in use is `<thing being formatted>As<how to format it>` (eg `InstIdAsType`, `TypeIdAsRawType`). Would it make sense to be consistent with that? Eg, `BoolAsSelect`, `IntAsSelect`

I understand the desire to namespace these a bit, especially since they're effectively global (in namespace `Carbon`), but I guess `FormatInt` doesn't convey to me that it's selecting a string format based on the int value. Maybe `FormatIntAsSelect` if you still want the namespacing?